doc/source/fq_nmod_sparse_mat.rst

```diff
@@ -0,0 +1,295 @@
+
+    Given a matrix ``M`` and target vector ``b``, use the Lanczos algorithm to
+    find a vector ``x`` such that Mx = b, returns `1` if successful and `0` if not
+
+.. function:: int fq_nmod_sparse_mat_solve_wiedemann(fq_nmod_struct *x, const fq_nmod_sparse_mat_t M, const fq_nmod_struct *b, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M`` and target vector ``b``, use the Wiedemann algorithm to
+    find a vector ``x`` such that Mx = b, returns `1` if successful and `0` if not
+
+.. function:: int fq_nmod_sparse_mat_solve_block_lanczos(fq_nmod_struct *x, const fq_nmod_sparse_mat_t M, const fq_nmod_struct *b, slong block_size, flint_rand_t state, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M`` and target vector ``b``, use Coppersmith's block Lanczos 
+    algorithm (with specified block size) to find a vector ``x`` such that Mx = b, 
+    returns `1` if successful and `0` if not
+
+.. function:: int fq_nmod_sparse_mat_solve_block_wiedemann(fq_nmod_struct *x, const fq_nmod_sparse_mat_t M, const fq_nmod_struct *b, slong block_size, flint_rand_t state, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M`` and target vector ``b``, use Coppersmith's block Wiedemann
+    algorithm (with specified block size) to find a vector ``x`` such that Mx = b, 
+    returns `1` if successful and `0` if not
+
+Nullvector and nullspace computation
+--------------------------------------------------------------------------------
+
+.. function:: int fq_nmod_sparse_mat_nullvector_lanczos(fq_nmod_struct *x, const fq_nmod_sparse_mat_t M, flint_rand_t state, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M``, use the Lanczos algorithm to find a nullvector ``x`` 
+    s.t. Mx = 0, returns `1` if successful and `0` if not
+
+.. function:: int fq_nmod_sparse_mat_nullvector_wiedemann(fq_nmod_struct *x, const fq_nmod_sparse_mat_t M, flint_rand_t state, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M``, use the Wiedemann algorithm to find a nullvector ``x`` 
+    s.t. Mx = 0, returns `1` if successful and `0` if not
+
+.. function:: int fq_nmod_sparse_mat_nullvector_block_lanczos(fq_nmod_struct *x, const fq_nmod_sparse_mat_t M, slong block_size, flint_rand_t state, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M``, use Coppersmith's block Lanczos algorithm to find a 
+    nullvector ``x`` s.t. Mx = 0, returns `1` if successful and `0` if not
+
+.. function:: int fq_nmod_sparse_mat_nullvector_block_wiedemann(fq_nmod_struct *x, const fq_nmod_sparse_mat_t M, slong block_size, flint_rand_t state, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M``, use Coppersmith's block Wiedemann algorithm to find a 
+    nullvector ``x`` s.t. Mx = 0, returns `1` if successful and `0` if not
+
+.. function:: int fq_nmod_sparse_mat_nullspace_rref(fq_nmod_mat_t X, const fq_nmod_sparse_mat_t M, const fq_nmod_ctx_t ctx)
+
+    Given a matrix ``M``, use the reduced row echelon form to construct the 
```

Comment:
Functions shouldn't initialise their outputs. It makes it hard to match up init's and clears. I know this is a major pain in the neck for nullspace, but the convention we have in fmpz_mat is to ask the user to supply a big enough matrix X which they have initialised. It's a controversial choice, but it is too late to change it now.